models/tshirt/include/Tshirt.h

```diff
@@ -0,0 +1,284 @@
+                const tshirt_state& state, const double activation, const double max_flow_velocity)
+        {
+            reservoirs.resize(params.nash_n);
+            for ( unsigned long i = 0; i < reservoirs.size(); ++i )
+            {
+                //construct a single outlet nonlinear reservoir
+                reservoirs[i] = Nonlinear_Reservoir(0, params.Ssmax, state.Snash[i], params.Kn, 1, activation,
+                        max_flow_velocity);
+            }
+        }
+
+        //! run one time step of tshirt
+        static int run(
+                double dt,
+                tshirt_params params,        //!< static parameters for tshirt
+                tshirt_state state,          //!< model state
+                tshirt_state& new_state,     //!< model state struct to hold new model state
+                tshirt_fluxes& fluxes,       //!< model flux object to hold calculated fluxes
+                double input_flux_meters,          //!< the amount water entering the system this time step
+                // TODO: should/can this be a smart pointer?
+                giuh_kernel* giuh_obj,       //!< kernel object for calculating GIUH runoff from subsurface lateral flow
+                void* et_params)            //!< parameters for the et function
+        {
+            double column_total_soil_moisture_deficit = params.Ssmax - state.Ss;
+
+            // Note this surface runoff value has not yet performed GIUH calculations
+            double surface_runoff, subsurface_infiltration_flux;
+
+            Schaake_partitioning_scheme(dt, calc_Cschaake(params), column_total_soil_moisture_deficit, input_flux_meters,
+                    &surface_runoff, &subsurface_infiltration_flux);
+
+            double Sfc = calc_Sfc(params, state);
+
+            vector<Reservoir_Outlet> subsurface_outlets;
+
+            // Keep track of the indexes of the specific outlets for later access
+            int lf_outlet_index = 0;
+            int perc_outlet_index = 1;
+
+            // TODO: confirm these are correct
+            // Max transmissivity
+            double max_lateral_flow = params.satdk * params.multiplier * params.Ssmax;
+            double max_perc_flow = params.satdk;
+
+            // init subsurface later flow outlet
+            subsurface_outlets[lf_outlet_index] = Reservoir_Outlet(params.Klf, 1.0, Sfc, max_lateral_flow);
+            // init subsurface percolation flow outlet
+            subsurface_outlets[perc_outlet_index] = Reservoir_Outlet(params.satdk * params.slope, 1.0, Sfc, max_perc_flow);
+
+            Nonlinear_Reservoir subsurface_reservoir(0.0, params.depth, state.Ss, subsurface_outlets);
+
+            double subsurface_excess;
+            subsurface_reservoir.response_meters_per_second(subsurface_infiltration_flux, dt, subsurface_excess);
+
+            // lateral subsurface flow
+            double Qlf = subsurface_reservoir.velocity_meters_per_second_for_outlet(lf_outlet_index);
```

Comment:
All of the usage of the subsurface reservoir look good to me up until this point. The only concern here is that velocity_meters_per_second_for_outlet uses the updated state of current_storage_height_meters after the response_meters_per_second is called. Therefore, the velocities that are returned reflect the end velocity after the drainage and not the velocity during the drainage. The velocity after drainage might actually be what is needed though. I am a unsure from the diagram and this implementation which velocity is actually needed.

If the velocity during the drainage is needed, I can create a small separate function that returns the velocities that were used in the drainage/response_meters_per_second, and the Reservoir_Outlet class would also maintain that as a state variable also.

Reply:
After further discussion, we concluded it is the velocity during the drainage. I am updating the nonlinear reservoir code to appropriately return this. @jdmattern-noaa